File: gto-support-jsonapi-retrofit2/src/main/java/org/ccci/gto/android/common/jsonapi/retrofit2/JsonApiConverterFactory.java
package org.ccci.gto.android.common.jsonapi.retrofit2;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.ccci.gto.android.common.jsonapi.JsonApiConverter;
import org.ccci.gto.android.common.jsonapi.JsonApiUtils;
import org.ccci.gto.android.common.jsonapi.model.JsonApiObject;
import org.ccci.gto.android.common.jsonapi.retrofit2.annotation.JsonApiInclude;
import org.json.JSONException;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

public class JsonApiConverterFactory extends Converter.Factory {
    static final MediaType MEDIA_TYPE = MediaType.parse(JsonApiObject.MEDIA_TYPE);

    @NonNull
    final JsonApiConverter mConverter;

    private JsonApiConverterFactory(@NonNull final JsonApiConverter converter) {
        mConverter = converter;
    }

    @NonNull
    public static JsonApiConverterFactory create(@NonNull final JsonApiConverter converter) {
        return new JsonApiConverterFactory(converter);
    }

    @NonNull
    public static JsonApiConverterFactory create(@NonNull final Class<?>... classes) {
        return new JsonApiConverterFactory(new JsonApiConverter.Builder().addClasses(classes).build());
    }

    @Nullable
    @Override
    public Converter<?, RequestBody> requestBodyConverter(@NonNull final Type type,
                                                          final Annotation[] parameterAnnotations,
                                                          final Annotation[] methodAnnotations,
                                                          final Retrofit retrofit) {
        // find any annotations we care about
        JsonApiInclude include = null;
        for (final Annotation ann : parameterAnnotations) {
            if (ann instanceof JsonApiInclude) {
                include = (JsonApiInclude) ann;
            }
        }

        final Class<?> c = JsonApiUtils.getRawType(type);
        if (JsonApiObject.class.isAssignableFrom(c)) {
            if (!(type instanceof ParameterizedType)) {
                throw new IllegalArgumentException("JsonApiObject needs to be parameterized");
            }

            return new JsonApiObjectRequestBodyConverter(include);
        } else if (List.class.isAssignableFrom(c)) {
            // TODO
        } else if (mConverter.supports(c)) {
            return new ObjectRequestBodyConverter(include);
        }
        return null;
    }

    @Nullable
    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(@NonNull final Type type, final Annotation[] annotations,
                                                            final Retrofit retrofit) {
        final Class<?> c = JsonApiUtils.getRawType(type);
        if (JsonApiObject.class.isAssignableFrom(c)) {
            if (!(type instanceof ParameterizedType)) {
                throw new IllegalArgumentException("JsonApiObject needs to be parameterized");
            }

            // find the data type
            final Class<?> dataType = JsonApiUtils.getRawType(((ParameterizedType) type).getActualTypeArguments()[0]);
            return new JsonApiObjectResponseBodyConverter<>(dataType);
        } else if (List.class.isAssignableFrom(c)) {
            // TODO
        } else if (mConverter.supports(c)) {
            return new ObjectResponseBodyConverter<>(c);
        }

        return null;
    }

    private class JsonApiObjectRequestBodyConverter implements Converter<JsonApiObject<?>, RequestBody> {
        @Nullable
        private final String[] mInclude;

        JsonApiObjectRequestBodyConverter(@Nullable final JsonApiInclude include) {
            mInclude = include != null ? include.value() : new String[0];
        }

        @Override
        public RequestBody convert(final JsonApiObject<?> value) throws IOException {
            return RequestBody.create(MEDIA_TYPE, mConverter.toJson(value, mInclude).getBytes("UTF-8"));
        }
    }

    private class JsonApiObjectResponseBodyConverter<T> implements Converter<ResponseBody, JsonApiObject<T>> {
        private final Class<T> mDataType;

        JsonApiObjectResponseBodyConverter(@NonNull final Class<T> type) {
            mDataType = type;
        }

        @Override
        public JsonApiObject<T> convert(final ResponseBody value) throws IOException {
            try {
                return mConverter.fromJson(value.string(), mDataType);
            } catch (final JSONException e) {
                throw new IOException("Error parsing JSON", e);
            }
        }
    }

    private class ObjectRequestBodyConverter implements Converter<Object, RequestBody> {
        @NonNull
        private final JsonApiObjectRequestBodyConverter mWrappedConverter;

        ObjectRequestBodyConverter(@Nullable final JsonApiInclude include) {
            mWrappedConverter = new JsonApiObjectRequestBodyConverter(include);
        }

        @Override
        public RequestBody convert(final Object value) throws IOException {
            return mWrappedConverter.convert(JsonApiObject.single(value));
        }
    }

    private class ObjectResponseBodyConverter<T> implements Converter<ResponseBody, T> {
        private final JsonApiObjectResponseBodyConverter<T> mWrappedConverter;

        ObjectResponseBodyConverter(@NonNull final Class<T> type) {
            mWrappedConverter = new JsonApiObjectResponseBodyConverter<>(type);
        }

        @Override
        public T convert(final ResponseBody value) throws IOException {
            return mWrappedConverter.convert(value).getDataSingle();
        }
    }
}

File: gto-support-jsonapi-retrofit2/src/main/kotlin/org/ccci/gto/android/common/jsonapi/retrofit2/JsonApiConverterFactory.kt
package org.ccci.gto.android.common.jsonapi.retrofit2

import java.io.IOException
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import okhttp3.MediaType
import okhttp3.RequestBody
import okhttp3.ResponseBody
import org.ccci.gto.android.common.jsonapi.JsonApiConverter
import org.ccci.gto.android.common.jsonapi.JsonApiUtils
import org.ccci.gto.android.common.jsonapi.model.JsonApiObject
import org.ccci.gto.android.common.jsonapi.retrofit2.annotation.JsonApiInclude
import org.ccci.gto.android.common.jsonapi.retrofit2.model.JsonApiRetrofitObject
import org.json.JSONException
import retrofit2.Converter
import retrofit2.Retrofit

private val MEDIA_TYPE = MediaType.parse(JsonApiObject.MEDIA_TYPE)

class JsonApiConverterFactory(private val converter: JsonApiConverter) : Converter.Factory() {
    constructor(vararg classes: Class<*>) : this(JsonApiConverter.Builder().addClasses(*classes).build())

    companion object {
        @Deprecated(
            "Since v4.0.1, use constructor instead.",
            ReplaceWith(
                "JsonApiConverterFactory(converter)",
                "org.ccci.gto.android.common.jsonapi.retrofit2.JsonApiConverterFactory",
            ),
        )
        fun create(converter: JsonApiConverter) = JsonApiConverterFactory(converter)

        @Deprecated(
            "Since v4.0.1, use constructor instead.",
            ReplaceWith(
                "JsonApiConverterFactory(*classes)",
                "org.ccci.gto.android.common.jsonapi.retrofit2.JsonApiConverterFactory",
            ),
        )
        fun create(vararg classes: Class<*>) = JsonApiConverterFactory(*classes)
    }

    override fun requestBodyConverter(
        type: Type,
        parameterAnnotations: Array<Annotation>,
        methodAnnotations: Array<Annotation>,
        retrofit: Retrofit,
    ): Converter<*, RequestBody>? {
        val rawType = JsonApiUtils.getRawType(type)
        val include = parameterAnnotations.filterIsInstance<JsonApiInclude>().firstOrNull()
        return when {
            JsonApiObject::class.java.isAssignableFrom(rawType) -> {
                require(type is ParameterizedType) { "JsonApiObject needs to be parameterized" }
                JsonApiObjectRequestBodyConverter(include)
            }
            Collection::class.java.isAssignableFrom(rawType) -> null // TODO
            converter.supports(rawType) -> ObjectRequestBodyConverter(include)
            else -> null
        }
    }

    override fun responseBodyConverter(
        type: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit,
    ): Converter<ResponseBody, *>? {
        val rawType = JsonApiUtils.getRawType(type)
        return when {
            JsonApiObject::class.java.isAssignableFrom(rawType) -> {
                require(type is ParameterizedType) { "JsonApiObject needs to be parameterized" }
                return JsonApiObjectResponseBodyConverter(JsonApiUtils.getRawType(type.actualTypeArguments[0]))
            }
            Collection::class.java.isAssignableFrom(rawType) -> null // TODO
            converter.supports(rawType) -> ObjectResponseBodyConverter(rawType)
            else -> null
        }
    }

    private inner class JsonApiObjectRequestBodyConverter(
        include: JsonApiInclude?,
    ) : Converter<JsonApiObject<*>, RequestBody> {
        private val options = JsonApiConverter.Options.builder()
            .apply {
                when {
                    include == null -> include()
                    include.all -> includeAll()
                    else -> include(*include.value)
                }
            }
            .build()

        override fun convert(value: JsonApiObject<*>): RequestBody {
            val options = if (value is JsonApiRetrofitObject<*>) options.merge(value.options) else options
            return RequestBody.create(MEDIA_TYPE, converter.toJson(value, options).toByteArray(Charsets.UTF_8))
        }
    }

    private inner class JsonApiObjectResponseBodyConverter<T>(
        private val type: Class<T>,
    ) : Converter<ResponseBody, JsonApiObject<T>> {
        override fun convert(value: ResponseBody) = try {
            converter.fromJson(value.string(), type)
        } catch (e: JSONException) {
            throw IOException("Error parsing JSON", e)
        }
    }

    private inner class ObjectRequestBodyConverter(include: JsonApiInclude?) : Converter<Any?, RequestBody> {
        private val wrappedConverter = JsonApiObjectRequestBodyConverter(include)
        override fun convert(value: Any?) = wrappedConverter.convert(JsonApiObject.single(value))
    }

    private inner class ObjectResponseBodyConverter<T>(type: Class<T>) : Converter<ResponseBody, T?> {
        private val wrappedConverter = JsonApiObjectResponseBodyConverter(type)
        override fun convert(value: ResponseBody) = wrappedConverter.convert(value).dataSingle
    }
}
